mistralrs-core/src/utils/model_config.rs

```diff
@@ -0,0 +1,328 @@
+use anyhow::Result;
+use candle_core::quantized::{gguf_file, ggml_file};
+use candle_nn::VarBuilder;
+use std::{collections::HashMap, path::PathBuf};
+use super::varbuilder_utils::{from_mmaped_safetensors, load_preload_adapters};
+
+use crate::{
+    DeviceMapMetadata,
+    lora::{Ordering, LoraConfig},
+    xlora_models::XLoraConfig,
+    pipeline::ModelPaths,
+};
+
+#[derive(derive_more::From)]
+pub struct FileGGML {
+    pub ct: ggml_file::Content,
+    pub gqa: usize,
+}
+
+#[derive(derive_more::From)]
+pub struct FileGGUF<'a> {
+    pub ct: gguf_file::Content,
+    pub reader: &'a mut std::fs::File,
+}
+
+#[derive(derive_more::From)]
+pub struct Device<'a> {
+    pub device: &'a candle_core::Device,
+    pub mapper: DeviceMapMetadata,
+}
+
+pub struct Adapter<'a> {
+    pub xlora_config: Option<XLoraConfig>,
+    pub lora_config: &'a [((String, String), LoraConfig)],
+    pub vb: VarBuilder<'a>,
+    pub ordering: &'a Ordering,
+    pub preload_adapters: Option<HashMap<String, (VarBuilder<'a>, LoraConfig)>>,
+}
+
+impl<'a> Adapter<'a> {
+    // NOTE: It is not possible to store references for values returned by: load_preload_adapters() + from_mmaped_safetensors(),
+    // As referenced value would drop after this method, Adapter takes ownership of vb + preload_adapters
+    // and then passes by reference to the `from_gguf()` / `from_ggml()` methods when proxying to params.
+    // NOTE: Due to reference usage persisting in returned struct, additional lifetime annotations were required.
+    pub fn try_new<'b: 'a>(
+        paths: &'b Box<dyn ModelPaths>,
+        device: &'b candle_core::Device,
+        silent: bool,
+        is_xlora: bool,
+    ) -> Result<Self> {
+        let lora_config = paths.get_adapter_configs().as_ref().unwrap();
+        let ordering = paths.get_ordering().as_ref().unwrap();
+        let preload_adapters = load_preload_adapters(
+            paths.get_lora_preload_adapter_info(),
+            candle_core::DType::F32,
+            device,
+            silent,
+        )?;
+
+        // X-LoRA support:
+        let mut xlora_paths: Vec<PathBuf> = vec![];
+        let mut xlora_config: Option<XLoraConfig> = None;
+        if is_xlora {
+            xlora_paths = vec![paths.get_classifier_path().as_ref().unwrap().to_path_buf()];
+            xlora_config = Some(paths.get_classifier_config().as_ref().unwrap().clone());
+        }
+
+        // Create VarBuilder:
+        // TODO: `from_mmaped_safetensors` has `xlora_paths` as the 2nd param, is this a bug?

```

Comment:
No, perhaps we should rename `xlora_paths` to `xlora_classifier_path` though as that is really what it carries here.

The 2nd param of `from_mmaped_safetensors` takes the adapters paths, perhaps we should rename that accordingly too?